### ISSUE TITLE
Android: grass texture problems (OpenGL ES 2 driver transparency bug)

### ISSUE THREAD

**Issue by nerzhul** [minetest/minetest#5600]
We have a very anoying bug with grass on Android a part of the texture is white
![screenshot_2017-04-16-09-46-03](https://cloud.githubusercontent.com/assets/119752/25069746/b2b2a8f2-228a-11e7-8a5e-9be3e90aa61b.png)

Tested with current minetest_game and 0.4.15

**Comment by Wayward1** [comment]
May be due to a recent change, I don't experience this on my current build (about a month old)

**Comment by Wayward1** [comment]
I just git pulled the latest engine and game, and I still don't experience this

![screenshot_2017-04-16-13-17-39](https://cloud.githubusercontent.com/assets/8821916/25072849/23e81414-22a7-11e7-9f12-8811733b2025.png)

**Comment by paramat** [comment]
Custom texture pack?

**Comment by nerzhul** [comment]
@paramat no it's minetest_game regular, i test in Android environment like release for play store

**Comment by MuhdNurHidayat** [comment]
that's problem with driver or renderer, I think...

because I actually experienced that when I change Minetest setting to force use ogles2 instead of default ogles1 while troubleshooting the problem on [Why is red colour missing from in-game screenshot?](https://forum.minetest.net/viewtopic.php?p=264038#p264038)

**Comment by MuhdNurHidayat** [comment]
I can't find button to edit comment on mobile page ver. of github so I write new one.

It is video driver, not renderer (but that's its job too).

The issue did not happen on ogles1 (OpenGL ES 1.0) ... My phone actually use OpenGL ES 3.0 (ogles3) but Minetest default it to ogles1 so the issue did not happen.

I came across it only when I change setting via minetest.conf to force it use ogles2 and ogles3 as I try to resolve that issue on [Android - Why is red colour missing from in-game screenshot?](https://forum.minetest.net/viewtopic.php?f=6&t=16872) . ogles2 seems to have the effect you stated, whereas ogles3 crashed the game.

The setting is at `Settings > Advanced Settings > Client > Graphics > In-Game > Advanced > Video driver` but there's no ogles2 or ogles3 from dropdown which is why I change it via minetest.conf
![setting](https://puu.sh/vnduU/7537133aa8.png "Setting page")

Is this same like your issue? This happen at ogles2 for me:
![ogles2](https://puu.sh/vne8N/737ae9fc28.jpg "ogles2")

The same position with default setting (ogles1) is okay for me:
![ogles1](https://puu.sh/vnetF/a1c6e8a56b.png "ogles1")

**Comment by Wayward1** [comment]
I can confirm, switching to ogles2 does indeed give the same results:

![screenshot_2017-04-28-21-39-24](https://cloud.githubusercontent.com/assets/8821916/25551931/5398fb48-2c5b-11e7-922f-72da1f43fb91.png)

It seems your device is somehow using the wrong video driver.

**Comment by Ferk** [comment]
I've experienced this as well in the past with some android devices, it's definitely a driver issue.

It's not just plantlike, it happens as well for torches and any billboard-like mesh.

**Comment by kilbith** [comment]
Duplicate of https://github.com/minetest/minetest/issues/2084

**Comment by Wayward1** [comment]
Idk, is it a duplicate if his device chose the wrong driver by default? We can't expect users to manually edit the .conf just to play.

**Comment by nerzhul** [comment]
@Wayward1 it's a vanilla build, i release mt on google play store, and i tested an intermediate build to look at players to database on android, i found a bug on progress bar, fixed now, but for this bug i didn't do anything. Editing .conf is unacceptable on android, yes

**Comment by paramat** [comment]
Reopening as this has better and newer discussion than #2084

**Comment by numberZero** [comment]
Just found out that GLES 2 **does not support alpha test at all.** Only alpha blending.
As a workaround, `EMT_TRANSPARENT_ALPHA_CHANNEL` could be used instead of `EMT_TRANSPARENT_ALPHA_CHANNEL_REF`; it’s slower but should work, at least.

**Comment by stujones11** [comment]
> it’s slower but should work, at least.

AFAIK this would suffer from the same depth sorting issues as using translucency, like not being visible from underwater.

**Comment by stujones11** [comment]
@numberZero Maybe this patch could help? https://github.com/freeminer/irrlicht/commit/854c90a378b9153f4143f595242f35de2beacb8a

**Comment by numberZero** [comment]
IrrLicht uses that shader for EMT_SOLID (EMT_TRANSPARENT_ALPHA_CHANNEL_REF is never ever referenced in the ogles2 driver). It uses MaterialTypeParam as alpha threshold. Try setting it to 0.5.

**Comment by stujones11** [comment]
> It uses MaterialTypeParam as alpha threshold. Try setting it to 0.5.

Thanks, works like a charm for me anyway :) Still, I get that (or very similar) effect when I enable basic shaders. However, this true for both ogles1 & ogles2 so maybe a separate issue.

**Comment by Wayward1** [comment]
Yes, shaders currently don't work on Android, as stated in #2060.

**Comment by stujones11** [comment]
Ah, didn't see that. Any chance you can test the above patch on your build?

**Comment by Wayward1** [comment]
I just did, it works great :)

**Comment by numberZero** [comment]
@stujones11 In the case you enable shaders, IrrLicht’s builtin shaders are not used. You need to update MT shaders then. Like this (may be needed for other shaders as well):
```diff
--- a/client/shaders/nodes_shader/opengl_fragment.glsl
+++ b/client/shaders/nodes_shader/opengl_fragment.glsl
@@ -180,6 +180,8 @@ void main(void)
 	}
 #endif
 	vec4 base = texture2D(baseTexture, uv).rgba;
+	if (base.a < 0.5)
+		discard;
 
 #ifdef ENABLE_BUMPMAPPING
 	if (use_normalmap) {
```

**Comment by stujones11** [comment]
Well that looks like it ought to work but has no effect whatsoever. In fact, none of the shaders seem to have any effect other than disabling irrlicht's built-in shaders. This makes me wonder if any shaders are loaded at all but I see nothing relevant in the logs.

Probably best sharders are disabled for android until fixed, however, that is another issue #2060

**Comment by paramat** [comment]
What's the status of this? What needs to be done?
It seems that shaders need to be either fixed or disabled for Android before 0.5 release.

**Comment by numberZero** [comment]
> This makes me wonder if any shaders are loaded at all

Oh, I just remembered one thing: GLES doesn’t need to support source-code shaders. It may only support precompiled ones. So I guess MT shaders are not loaded at all, and IrrLicht uses some fallback shader then (GLES 2+ can’t render without a shader).

**Comment by paramat** [comment]
So we should disable shaders on Android?

**Comment by numberZero** [comment]
> So we should disable shaders on Android?

Unless we can compile them to a format Android supports. Sorry but I don’t know much about that, I don’t use Android myself.

**Comment by paramat** [comment]
Seems best to disable them. 0.5 needs to sort out Android so the shaders options should be removed to avoid a bad impression.

**Comment by stujones11** [comment]
AFAIK this particular issue was resolved by #6971
